solidity/ecdsa/contracts/libraries/DKG.sol

```diff
@@ -0,0 +1,509 @@
+    struct Result {
+        // Claimed submitter candidate group member index.
+        // Must be in range [1, groupSize].
+        uint256 submitterMemberIndex;
+        // Generated candidate group public key
+        bytes groupPubKey;
+        // Array of misbehaved members indices (disqualified or inactive).
+        // Indices must be in range [1, groupSize], unique, and sorted in ascending
+        // order.
+        uint8[] misbehavedMembersIndices;
+        // Concatenation of signatures from members supporting the result.
+        // The message to be signed by each member is keccak256 hash of the
+        // calculated group public key, misbehaved members indices and DKG
+        // start block. The calculated hash should be prefixed with prefixed with
+        // `\x19Ethereum signed message:\n` before signing, so the message to
+        // sign is:
+        // `\x19Ethereum signed message:\n${keccak256(
+        //    groupPubKey, misbehavedMembersIndices, dkgStartBlock
+        // )}`
+        bytes signatures;
+        // Indices of members corresponding to each signature. Indices must be
+        // be in range [1, groupSize], unique, and sorted in ascending order.
+        uint256[] signingMembersIndices;
+        // Identifiers of candidate group members as outputted by the group
+        // selection protocol.
+        uint32[] members;
+        // Keccak256 hash of group members identifiers that actively took part
+        // in DKG (excluding IA/DQ members).
+        bytes32 membersHash;
+    }
+
+    /// @notice States for phases of group creation. The states doesn't include
+    ///         timeouts which should be tracked and notified individually.
+    enum State {
+        // Group creation is not in progress. It is a state set after group creation
+        // completion either by timeout or by a result approval.
+        IDLE,
+        // Group creation is awaiting the seed and sortition pool is locked.
+        AWAITING_SEED,
+        // Off-chain DKG protocol execution is in progress. A result is being calculated
+        // by the clients in this state. It's not yet possible to submit the result.
+        KEY_GENERATION,
+        // After off-chain DKG protocol execution the contract awaits result submission.
+        // This is a state to which group creation returns in case of a result
+        // challenge notification.
+        AWAITING_RESULT,
+        // DKG result was submitted and awaits an approval or a challenge. If a result
+        // gets challenge the state returns to `AWAITING_RESULT`. If a result gets
+        // approval the state changes to `IDLE`.
+        CHALLENGE
+    }
+
+    /// @dev Size of a group in the threshold relay.
+    uint256 public constant groupSize = 100;
+
+    /// @notice Time in blocks after which DKG result is complete and ready to be
+    //          published by clients.
+    uint256 public constant offchainDkgTime = 5 * (1 + 5) + 2 * (1 + 10) + 20;
```

Comment:
One idea of how we could solve it: have a governable value for `offchainDkgTime` and make member 1 always eligible to submit. Someone needs to be eligible in case it takes less time for DKG than expected on-chain. This is possible given we may need to retry the protocol several times to have a group with 100% operator availability at the time of DKG. Remember that inactive operators are not well supported for the off-chain tECDSA key-gen protocol we currently use.

Reply:
> make member 1 always eligible to submit.

I think it could be the case for random beacon as well if it makes things simpler. Just please think more about it first, I haven't had a chance to spend on it enough time to ensure it's not opening the system for some kind of an attack.

Reply:
I removed submission eligibility queue enforcement in 00a413f92c3a4f2151b803705521430af46435e4

I also added comment to `offchainDkgTime` that it needs an update (2c825f5db31fa9409a15ca0e9d361a6b1425fcab).